tools/samtools.wdl

```diff
@@ -327,3 +326,263 @@ task addreplacerg {
+        prefix: "Prefix for the collated BAM and FastQ files. The extensions `.collated.bam` and `[,_R1,_R2,.singleton].fastq.gz` will be added."
+        f: "Only output alignments with all bits set in INT present in the FLAG field. INT can be specified in hex by beginning with `0x` (i.e. /^0x[0-9A-F]+/) or in octal by beginning with `0` (i.e. /^0[0-7]+/)."
+        F: "Do not output alignments with any bits set in INT present in the FLAG field. INT can be specified in hex by beginning with `0x` (i.e. /^0x[0-9A-F]+/) or in octal by beginning with `0` (i.e. /^0[0-7]+/). This defaults to 0x900 representing filtering of secondary and supplementary alignments."
+        G: "Only EXCLUDE reads with all of the bits set in INT present in the FLAG field. INT can be specified in hex by beginning with `0x` (i.e. /^0x[0-9A-F]+/) or in octal by beginning with `0` (i.e. /^0[0-7]+/)."
+        fast_mode: "Fast mode for `samtools collate` (primary alignments only)"
+        store_collated_bam: "Save the collated BAM (true) or delete it after FastQ split (false)?"
+        paired_end: "Is the data paired-end?"
+        interleaved: "Create an interleaved FastQ file from paired-end data?"
+        output_singletons: "Output singleton reads as their own FastQ?"
+        use_all_cores: "Use all cores? Recommended for cloud environments. Not recommended for cluster environments."
+        modify_memory_gb: "Add to or subtract from dynamic memory allocation. Default memory is determined by the size of the inputs. Specified in GB."
+        modify_disk_size_gb: "Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB."
+        ncpu: "Number of cores to allocate for task"
+        max_retries: "Number of times to retry in case of failure"
+    }
+
+    input {
+        File bam
+        String prefix = basename(bam, ".bam")
+        String f = "0"
+        String F = "0x900"
+        String G = "0"
+        Boolean fast_mode = true
+        Boolean store_collated_bam = false
+        Boolean paired_end = true
+        Boolean interleaved = false
+        Boolean output_singletons = false
+        Boolean use_all_cores = false
+        Int modify_memory_gb = 0
+        Int modify_disk_size_gb = 0
+        Int ncpu = 1
+        Int max_retries = 1
+    }
+
+    Float bam_size = size(bam, "GiB")
+    Int memory_gb = ceil(bam_size * 0.2) + 4 + modify_memory_gb
+    Int disk_size_gb = ceil((bam_size * 5) + 10) + modify_disk_size_gb
+
+    command <<<
+        set -euo pipefail
+
+        n_cores=~{ncpu}
+        if [ "~{use_all_cores}" = "true" ]; then
+            n_cores=$(nproc)
+        fi
+
+        samtools collate \
+            --threads "$n_cores" \
+            ~{if fast_mode then "-f" else ""} \
+            # skip compression (and decompression) if not storing the output
+            ~{if store_collated_bam then "" else "-u"} \
```

Comment:
I thought we weren't using the uncompressed BAM output. However looking at this, since we're not storing it, if the subsequent `samtools fastq` call accepts it without error, I'm fine using it here.

Reply:
I'm about to submit jobs verifying this works (and seeing if it has any effect on runtime). If it's neutral on runtime I plan to remove it for simplicity's sake